packages/types-core/src/project/modulars/types.ts

```diff
@@ -0,0 +1,10 @@
+// Copyright 2020-2024 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import {BaseDataSource, IProjectManifest} from '../index';
+
+export interface INetworkCommonModule<D extends BaseDataSource = BaseDataSource> {
+  parseProjectManifest(raw: unknown): IProjectManifest<D>;
+  isRuntimeDs(ds: D): boolean;
+  isCustomDs(ds: D): boolean;
```

Comment:
Returning booleans mean we don't get the type narrowing. I think this might work.

```suggestion
  isRuntimeDs<DS extends D = D>(ds: D): ds is DS;
  isCustomDs<CDS extends BaseCustomDataSource = BaseCustomDataSource>(ds: D): ds is CDS;
```